File: cpp/2015/01/main.cpp
#include <iostream>
#include <fstream>
#include <streambuf>

template<typename T = char>
int last_floor(std::ifstream& file) {
  int floor{};
  for (std::istreambuf_iterator<T> it{file}, end; it != end; it++) {
    switch (*it) {
      case '(':
        ++floor;
        break;
      case ')':
        --floor;
        break;
    }
  }
  return floor;
}

template<typename T = char>
std::optional<int> first_cross(std::ifstream& file) {
  file.seekg(0);
  int count{};
  int floor{};
  for (std::istreambuf_iterator<T> it{file}, end; it != end; it++) {
    switch (*it) {
      case '(':
        ++floor;
        break;
      case ')':
        --floor;
        break;
    }
    ++count;
    if (floor < 0) return count;
  }
  return std::nullopt;
}

int main() {
  std::ifstream file("data.txt");

  const int ans1 = last_floor<char>(file);
  std::cout << "Answer 1: " << ans1 << std::endl;

  const int ans2 = first_cross(file).value_or(-1);
  std::cout << "Answer 2: " << ans2 << std::endl;

  return 0;
}

File: zig/2015/01/src/main.zig
const std = @import("std");

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const file = try std.fs.cwd().openFile("data.txt", .{});
    const data = try file.readToEndAlloc(allocator, 1 << 13);

    const ans1 = try last_floor(data);
    const ans2 = try first_cross(data);

    std.debug.print("Ans1: {d}\n", .{ans1});
    std.debug.print("Ans2: {d}\n", .{ans2.?});
}

fn last_floor(buf: []u8) !i32 {
    var floor: i32 = 0;
    for (buf) |c| switch (c) {
        '(' => floor += 1,
        ')' => floor -= 1,
        else => return error.UnknownCharacter,
    };

    return floor;
}

fn first_cross(buf: []u8) !?i32 {
    var floor: i32 = 0;
    for (buf, 1..) |c, i| {
        switch (c) {
            '(' => floor += 1,
            ')' => floor -= 1,
            else => return error.UnknownCharacter,
        }

        if (floor == -1) return @intCast(i);
    }

    return null;
}
